### PR TITLE
planner: change the stage of count star rewriter rule from logical to post

### LINKED ISSUE #37165: planner: optimize the performance of `SELECT count(*)`
## Enhancement
```mysql
mysql > explain analyze select count(*) from github_events;
+------------------------------+---------------+------------+--------------+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
| id                           | estRows       | actRows    | task         | access object       | execution info                                                                                                                                                                                                                                                                                                                                                            | operator info                             | memory  | disk |
+------------------------------+---------------+------------+--------------+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
| HashAgg_92                   | 1.00          | 1          | root         |                     | time:3.59s, loops:2, partial_worker:{wall_time:3.586945619s, concurrency:5, task_num:1, tot_wait:17.934327667s, tot_exec:13.429µs, tot_time:17.934350874s, max:3.586886692s, p95:3.586886692s}, final_worker:{wall_time:3.586962598s, concurrency:5, task_num:1, tot_wait:17.934505317s, tot_exec:19.739µs, tot_time:17.934528686s, max:3.586912325s, p95:3.586912325s}   | funcs:count(Column#83)->Column#33         | 9.86 KB | N/A  |
| └─TableReader_94             | 1.00          | 4          | root         | partition:all       | time:3.59s, loops:2, cop_task: {num: 6, max: 0s, min: 0s, avg: 0s, p95: 0s, copr_cache_hit_ratio: 0.00}                                                                                                                                                                                                                                                                   | data:ExchangeSender_93                    | N/A     | N/A  |
|   └─ExchangeSender_93        | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:3.57s, min:3.21s, avg: 3.37s, p80:3.57s, p95:3.57s, iters:4, tasks:4, threads:88}                                                                                                                                                                                                                                                                  | ExchangeType: PassThrough                 | N/A     | N/A  |
|     └─HashAgg_8              | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:3.57s, min:3.21s, avg: 3.37s, p80:3.57s, p95:3.57s, iters:4, tasks:4, threads:4}                                                                                                                                                                                                                                                                   | funcs:count(1)->Column#83                 | N/A     | N/A  |
|       └─TableFullScan_91     | 4897975649.00 | 4907282882 | mpp[tiflash] | table:github_events | tiflash_task:{proc max:3.37s, min:3.07s, avg: 3.21s, p80:3.37s, p95:3.37s, iters:78475, tasks:4, threads:88}                                                                                                                                                                                                                                                              | keep order:false, PartitionTableScan:true | N/A     | N/A  |
+------------------------------+---------------+------------+--------------+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
5 rows in set (3.63 sec)

mysql > explain analyze select count(actor_id) from github_events;
+------------------------------+---------------+------------+--------------+---------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------+---------+------+
| id                           | estRows       | actRows    | task         | access object       | execution info                                                                                                                                                                                                                                                                                                                                                           | operator info                                                | memory  | disk |
+------------------------------+---------------+------------+--------------+---------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------+---------+------+
| HashAgg_23                   | 1.00          | 1          | root         |                     | time:2.83s, loops:2, partial_worker:{wall_time:2.830765746s, concurrency:5, task_num:1, tot_wait:14.153378614s, tot_exec:5.102µs, tot_time:14.153392452s, max:2.830691203s, p95:2.830691203s}, final_worker:{wall_time:2.830816995s, concurrency:5, task_num:1, tot_wait:14.153613809s, tot_exec:60.547µs, tot_time:14.153677414s, max:2.830743316s, p95:2.830743316s}   | funcs:count(Column#35)->Column#33                            | 9.86 KB | N/A  |
| └─TableReader_25             | 1.00          | 4          | root         | partition:all       | time:2.83s, loops:2, cop_task: {num: 5, max: 0s, min: 0s, avg: 0s, p95: 0s, copr_cache_hit_ratio: 0.00}                                                                                                                                                                                                                                                                  | data:ExchangeSender_24                                       | N/A     | N/A  |
|   └─ExchangeSender_24        | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:2.81s, min:2.24s, avg: 2.56s, p80:2.81s, p95:2.81s, iters:4, tasks:4, threads:88}                                                                                                                                                                                                                                                                 | ExchangeType: PassThrough                                    | N/A     | N/A  |
|     └─HashAgg_8              | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:2.81s, min:2.24s, avg: 2.56s, p80:2.81s, p95:2.81s, iters:4, tasks:4, threads:4}                                                                                                                                                                                                                                                                  | funcs:count(gharchive_dev.github_events.actor_id)->Column#35 | N/A     | N/A  |
|       └─TableFullScan_22     | 4897975649.00 | 4907282882 | mpp[tiflash] | table:github_events | tiflash_task:{proc max:2.78s, min:2.22s, avg: 2.53s, p80:2.78s, p95:2.78s, iters:80546, tasks:4, threads:88}                                                                                                                                                                                                                                                             | keep order:false, PartitionTableScan:true                    | N/A     | N/A  |
+------------------------------+---------------+------------+--------------+---------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------+---------+------+
5 rows in set (2.88 sec)
```
Now we scan the primary key column in default. We can improve the performance of `SELECT count(*)` by scanning the del_tag column rather than the pk column.

BTW, the following case will be improved, too.
```mysql
mysql> explain analyze select count(*) from (select actor_id from github_events) as a;
+------------------------------+---------------+------------+--------------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
| id                           | estRows       | actRows    | task         | access object       | execution info                                                                                                                                                                                                                                                                                                                                                          | operator info                             | memory  | disk |
+------------------------------+---------------+------------+--------------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
| HashAgg_93                   | 1.00          | 1          | root         |                     | time:3.67s, loops:2, partial_worker:{wall_time:3.671883051s, concurrency:5, task_num:1, tot_wait:18.359093871s, tot_exec:5.013µs, tot_time:18.359105489s, max:3.671827322s, p95:3.671827322s}, final_worker:{wall_time:3.67190697s, concurrency:5, task_num:1, tot_wait:18.359126767s, tot_exec:27.704µs, tot_time:18.359157552s, max:3.671852045s, p95:3.671852045s}   | funcs:count(Column#83)->Column#33         | 9.86 KB | N/A  |
| └─TableReader_95             | 1.00          | 4          | root         | partition:all       | time:3.67s, loops:2, cop_task: {num: 7, max: 0s, min: 0s, avg: 0s, p95: 0s, copr_cache_hit_ratio: 0.00}                                                                                                                                                                                                                                                                 | data:ExchangeSender_94                    | N/A     | N/A  |
|   └─ExchangeSender_94        | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:3.65s, min:3.25s, avg: 3.4s, p80:3.65s, p95:3.65s, iters:4, tasks:4, threads:88}                                                                                                                                                                                                                                                                 | ExchangeType: PassThrough                 | N/A     | N/A  |
|     └─HashAgg_9              | 1.00          | 4          | mpp[tiflash] |                     | tiflash_task:{proc max:3.65s, min:3.25s, avg: 3.4s, p80:3.65s, p95:3.65s, iters:4, tasks:4, threads:4}                                                                                                                                                                                                                                                                  | funcs:count(1)->Column#83                 | N/A     | N/A  |
|       └─TableFullScan_92     | 4898336416.00 | 4907409773 | mpp[tiflash] | table:github_events | tiflash_task:{proc max:3.43s, min:3.12s, avg: 3.23s, p80:3.43s, p95:3.43s, iters:78512, tasks:4, threads:88}                                                                                                                                                                                                                                                            | keep order:false, PartitionTableScan:true | N/A     | N/A  |
+------------------------------+---------------+------------+--------------+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------+---------+------+
```

### PR DESCRIPTION
**Pull request by elsa0520**
### What problem does this PR solve?

Issue Number: close #37165

Problem Summary:

### What is changed and how it works?

Change count star rewriter from logical rule to post phase

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/optimizer.go [hunk 1]
```diff
@@ -533,6 +534,119 @@ func prunePhysicalColumnsInternal(sctx sessionctx.Context, plan PhysicalPlan) er
...
+	}
+	physicalTableScan, ok := physicalAgg.Children()[0].(*PhysicalTableScan)
+	if !ok || !physicalTableScan.isFullScan() || physicalTableScan.StoreType != kv.TiFlash || len(physicalTableScan.schema.Columns) != 1 {
+		return
+	}
+	// rewrite datasource and agg args
+	rewriteTableScanAndAggArgs(physicalTableScan, physicalAgg.AggFuncs)
+}
+
+// Pick the narrowest and not null column from table
```
Comment (winoros) [inline planner/core/optimizer.go:608]: The comments are better began with `funcName picks ....`
This is the static check when the function is exported.

Comment (elsa0520) [reply planner/core/optimizer.go]: sure

File: planner/core/optimizer.go [hunk 2]
```diff
@@ -533,6 +534,119 @@ func prunePhysicalColumnsInternal(sctx sessionctx.Context, plan PhysicalPlan) er
...
+	for _, aggFunc := range aggFuncs {
+		constExpr, ok := aggFunc.Args[0].(*expression.Constant)
+		if !ok {
+			return
+		}
+		// count(null) shouldn't be rewritten
+		if constExpr.Value.IsNull() {
+			continue
+		}
+		aggFunc.Args[0] = resultColumn
```
Comment (winoros) [inline planner/core/optimizer.go:646]:
```suggestion
		aggFunc.Args[0] = resultColumn.Clone()
```

the information of the execution is bound in the expression.Column.(expression.Column.Index).
So we need to clone it here for safety.

(we need to set `aggFunc.Args[0].(*expression.Column).Index = 0` actually)

**Review by winoros — APPROVED** [review]

**Review by fixdb — APPROVED** [review]
+1

File: planner/core/optimizer.go [hunk 3]
```diff
@@ -533,6 +534,120 @@ func prunePhysicalColumnsInternal(sctx sessionctx.Context, plan PhysicalPlan) er
...
+	count(*) -> count(not null column)
+
+**Only for TiFlash**
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(constant)
+
+Pattern:
+PhysicalAggregation: count(constant)
+
```
Comment (bb7133) [inline planner/core/optimizer.go:550]: Why there is a blank line? Is it expected?

Comment (fixdb) [reply planner/core/optimizer.go:550]: Maybe a typo, we can delete it in later commits.

**Pull request merged**